### PR TITLE
[PLGN-101] InsightVM - Add Action Update Shared Credential

### PR DESCRIPTION
**Pull request by TKuzas-R7**
### Description

Describe the proposed changes:

  - Add shared Credentials ✅  
  - Add unit test ✅ 

### Notes for self
- double check plugin spec yaml ✅ 
- add actual examples for example input in help.md ✅ 
## PR Requirements

Developers, verify you have completed the following items by checking them off:

#### Unit Tests

Review our documentation on [generating](https://docs.rapid7.com/insightconnect/unit-test-generation) and [writing](https://docs.rapid7.com/insightconnect/unit-test-primer) plugin unit tests

- [x] Unit tests written for any new or updated code

#### In-Product Tests

If you are an InsightConnect customer or have access to an InsightConnect instance, the following in-product tests should be done:

- [ ] Screenshot of job output with the plugin changes
- [ ] Screenshot of the changed connection, actions, or triggers input within the InsightConnect workflow builder

### Style

Review the [style guide](https://docs.rapid7.com/insightconnect/style-guide/)

- [ ] For dependencies, pin [OS package](https://docs.rapid7.com/insightconnect/style-guide/#dockerfile) and [Python package](https://docs.rapid7.com/insightconnect/style-guide/#requirements.txt) versions
- [ ] For security, set least privileged account with ``USER nobody`` in the ``Dockerfile`` when possible
- [ ] For size, use the [slim SDK images](https://docs.rapid7.com/insightconnect/sdk-guide/#sdk-guide) when possible: ``komand/python-3-37-slim-plugin`` and ``komand/python-3-37-plugin``
- [ ] For error handling, use of [PluginException](https://docs.rapid7.com/insightconnect/error-handling-in-integrations/#plugin-exceptions) and [ConnectionTestException](https://docs.rapid7.com/insightconnect/error-handling-in-integrations#connection-exceptions)
- [ ] For logging, use [self.logger](https://docs.rapid7.com/insightconnect/sdk-guide/#logging)
- [ ] For docs, use [changelog style](https://docs.rapid7.com/insightconnect/style-guide/#changelog)
- [ ] For docs, validate markdown with ``make validate`` which calls ``mdl`` to lint ``help.md``

### Functional Checklist
- [ ] Work fully completed
- [ ] Functional
  - [ ] Any new actions/triggers include JSON [test files](https://docs.rapid7.com/insightconnect/style-guide/#tests) in the `tests/` directory created with `icon-plugin run -c sample $action > tests/$action.json`
  - [ ] Tests should all pass unless it's a negative test. Negative tests have a naming convention of `tests/$action_bad.json`
  - [ ] Unsuccessful tests should fail by raising an exception causing the plugin to die and an object should be returned on successful test
  - [ ] Add functioning test results to PR, sanitize any output if necessary
    * Single action/trigger `icon-plugin run -T tests/example.json --debug --jq`
    * All actions/triggers shortcut `icon-plugin run -T all --debug --jq` (use PR format at end)
  - [ ] Add functioning run results to PR, sanitize any output if necessary
    * Single action/trigger `icon-plugin run -R tests/example.json --debug --jq`
    * All actions/triggers shortcut `icon-plugin run -R all --debug --jq` (use PR format at end)

### Assessment

You must validate your work to reviewers:

1. Run `make validate` and make sure everything passes
2. Run the assessment tool: `icon-plugin run -A -R all -T all`. For single action validation: `icon-plugin run -A -R tests/my_action.json -T tests/my_action.json`
3. Copy (`icon-plugin ... | pbcopy`) and paste the output in **a new post** on this PR
4. Add required screenshots from the In-Product Tests section

### CODE REVIEW

File: plugins/rapid7_insightvm/.CHECKSUM [hunk 1]
```diff
@@ -366,6 +370,42 @@
 		{
 			"identifier": "new_scans/schema.py",
 			"hash": "e1e19bb26c002cd43bdad5dcf50a0aee"
+		},
```
Comment (cmcnally-r7) [inline plugins/rapid7_insightvm/.CHECKSUM:373]: You need to get rid of this stuff - I think this has happened you before

File: plugins/rapid7_insightvm/help.md [hunk 2]
```diff
@@ -1,4 +1,4 @@
-# Description
+ # Description
```
Comment (cmcnally-r7) [inline plugins/rapid7_insightvm/help.md:1]: Remove whitespace
```suggestion
# Description
```

File: plugins/rapid7_insightvm/help.md [hunk 3]
```diff
@@ -42,6 +42,58 @@ Example input:
...
+}
+```
+
+##### Output
+
+|Name|Type|Required|Description|
+|----|----|--------|-----------|
+|links|[]link|False|Hypermedia links to corresponding or related resources|
+
+
```
Comment (cmcnally-r7) [inline plugins/rapid7_insightvm/help.md:96]: You'll need example output here

File: plugins/rapid7_insightvm/plugin.spec.yaml [hunk 4]
```diff
@@ -1732,15 +1732,161 @@ types:
       required: true
   vulnerability_solution:
     links:
-      description: Hypermedia links to corresponding or related resources
       title: Links
+      description: Hypermedia links to corresponding or related resources
```
Comment (cmcnally-r7) [inline plugins/rapid7_insightvm/plugin.spec.yaml:1736]: good catch

**Review by cmcnally-r7 — COMMENTED** [review]
some small things

File: plugins/rapid7_insightvm/komand_rapid7_insightvm/actions/update_shared_credential/action.py [hunk 5]
```diff
@@ -0,0 +1,171 @@
...
+from ...util import endpoints
+from ...util.resource_requests import ResourceRequests
+from ...util.util import check_not_null, check_in_enum, make_payload
+
+
+def as400_cifs_cvs(account: dict):
+    service = account.get("service")
+    domain = account.get("domain", "")
+    username = check_not_null(account.get("username"))
+    password = check_not_null(account.get("password"))
```
Comment (cmcnally-r7) [inline plugins/rapid7_insightvm/komand_rapid7_insightvm/actions/update_shared_credential/action.py:13]: Username & Password seem to be very commonly occurring fields between functions.
Maybe we could do something like
```
def user_and_pw(username, pw):
    if username and pw:
        return check_not_null(account.get("username")), check_not_null(account.get("password"))
    else:
        return check_not_null(account.get("username"))

**Review by cmcnally-r7 — COMMENTED** [review]
.

File: plugins/rapid7_insightvm/komand_rapid7_insightvm/actions/update_shared_credential/action.py (hunk 5 above)
Comment (cmcnally-r7) [reply plugins/rapid7_insightvm/komand_rapid7_insightvm/actions/update_shared_credential/action.py:13]: And then retrieve the variables like this
```
(username, password) = user_and_pw(username, pw)
```

File: plugins/rapid7_insightvm/komand_rapid7_insightvm/actions/update_shared_credential/action.py [hunk 6]
```diff
@@ -0,0 +1,171 @@
+import insightconnect_plugin_runtime
+from insightconnect_plugin_runtime.exceptions import PluginException
+from .schema import UpdateSharedCredentialInput, UpdateSharedCredentialOutput, Input, Output, Component
+from ...util import endpoints
+from ...util.resource_requests import ResourceRequests
+from ...util.util import check_not_null, check_in_enum, make_payload
+
+
+def as400_cifs_cvs(account: dict):
```
Comment (cmcnally-r7) [inline plugins/rapid7_insightvm/komand_rapid7_insightvm/actions/update_shared_credential/action.py:9]:
```suggestion
def as400_cifs_cvs(account: dict) -> dict:
```

**Review by cmcnally-r7 — COMMENTED** [review]
Type hinting

File: plugins/rapid7_insightvm/komand_rapid7_insightvm/actions/update_shared_credential/action.py (hunk 5 above)
Comment (TKuzas-R7) [reply plugins/rapid7_insightvm/komand_rapid7_insightvm/actions/update_shared_credential/action.py:13]: smort, i will do exactly that

File: plugins/rapid7_insightvm/komand_rapid7_insightvm/util/endpoints.py [hunk 7]
```diff
@@ -316,6 +316,19 @@ def site_excluded_asset_groups(console_url, site_id):
         return f"{console_url}/api/3/sites/{site_id}/excluded_asset_groups"
 
 
+class SharedCredential:
+
+    @staticmethod
+    def update_shared_credential(console_url, cred_id):
```
Comment (cmcnally-r7) [inline plugins/rapid7_insightvm/komand_rapid7_insightvm/util/endpoints.py:322]:
```suggestion
    def update_shared_credential(console_url: str, cred_id: int(?)) -> str:
```

File: plugins/rapid7_insightvm/komand_rapid7_insightvm/util/update_shared_credential_util.py [hunk 8]
```diff
@@ -0,0 +1,117 @@
+from insightconnect_plugin_runtime.exceptions import PluginException
+
+from komand_rapid7_insightvm.util.util import check_in_enum, check_not_null
+
+
+def ssh_setup(account: dict):
```
Comment (cmcnally-r7) [inline plugins/rapid7_insightvm/komand_rapid7_insightvm/util/update_shared_credential_util.py:6]:
```suggestion
def ssh_setup(account: dict) -> Tuple[whatever, whatever, whatever]:
```

File: plugins/rapid7_insightvm/komand_rapid7_insightvm/util/update_shared_credential_util.py [hunk 9]
```diff
@@ -0,0 +1,117 @@
...
+    return {"service": service, "username": username, "password": password}
+
+
+def http(account: dict, service: str) -> dict:
+    realm = account.get("realm", "")
+    username, password = usr_and_pass(account)
+    return {"service": service, "realm": realm, "username": username, "password": password}
+
+
+def ms_sql_sybase(account: dict, service: str) -> dict:
```
Comment (cmcnally-r7) [inline plugins/rapid7_insightvm/komand_rapid7_insightvm/util/update_shared_credential_util.py:53]: I think this is considered one or the other. Union is _one or the other_
```suggestion
def ms_sql_sybase(account: dict, service: str) -> Union[dict, dict]:
```

**Review by cmcnally-r7 — COMMENTED** [review]
Looking good boss

File: plugins/rapid7_insightvm/komand_rapid7_insightvm/actions/update_shared_credential/action.py [hunk 10]
```diff
@@ -0,0 +1,48 @@
...
+        port_restriction = params.get("port_restriction", "")
+        site_assignment = check_not_null(params, "site_assignment")
+        check_in_enum(site_assignment, "site_assignment", ["all-sites", "specific-sites"])
+        if site_assignment == "all-sites":
+            sites = None
+        else:
+            sites = check_not_null(params, "sites")
+
+        service = check_not_null(account, "service")
+        service_dict = {"as400": as400_cifs_cvs, "cifs": as400_cifs_cvs, "cvs": as400_cifs_cvs, "cifshash": cifshash,
```
Comment (cmcnally-r7) [inline plugins/rapid7_insightvm/komand_rapid7_insightvm/actions/update_shared_credential/action.py:36]: Could me move this to a util file?

File: plugins/rapid7_insightvm/.CHECKSUM [hunk 11]
```diff
@@ -366,6 +370,42 @@
...
+			"identifier": "hello_trigger/schema.py",
+			"hash": "6fda62f9b28a899e2ecb7aab1311bbb2"
+		},
+		{
+			"identifier": "return_bad_json_trigger/schema.py",
+			"hash": "9a207816ba5b0ab0f9e4f40d86c4cc3e"
+		},
+		{
+			"identifier": "throw_exception_trigger/schema.py",
+			"hash": "23e4aee31e1c532d9ed50830efad1b8d"
```
Comment (cmcnally-r7) [inline plugins/rapid7_insightvm/.CHECKSUM:408]: do the thing

File: plugins/rapid7_insightvm/unit_test/test_update_shared_credential.py [hunk 12]
```diff
@@ -0,0 +1,716 @@
...
+
+@patch("requests.sessions.Session.put", side_effect=Util.mocked_requests)
+class TestUpdateSharedCredential(TestCase):
+    @classmethod
+    def setUpClass(cls) -> None:
+        cls.action = Util.default_connector(UpdateSharedCredential())
+
+        # make connector in util
+        # rename to USC
+        # add parameterized test for each service
```
Comment (cmcnally-r7) [inline plugins/rapid7_insightvm/unit_test/test_update_shared_credential.py:25]: Remove the comments

**Review by cmcnally-r7 — CHANGES_REQUESTED** [review]

File: plugins/rapid7_insightvm/komand_rapid7_insightvm/actions/update_shared_credential/action.py (hunk 10 above)
Comment (TKuzas-R7) [reply plugins/rapid7_insightvm/komand_rapid7_insightvm/actions/update_shared_credential/action.py:36]: we can move it all to the util file :)

File: plugins/rapid7_insightvm/komand_rapid7_insightvm/actions/update_shared_credential/action.py [hunk 13]
```diff
@@ -0,0 +1,29 @@
...
+        resource_helper = ResourceRequests(self.connection.session, self.logger)
+        endpoint = endpoints.SharedCredential.update_shared_credential(
+            self.connection.console_url, params.get(Input.ID)
+        )
+        account = params.get("account")
+        service = check_not_null(account, "service")
+        account_input = get_account_input(account, service)
+        payload = make_payload(params, account_input)
+        response = resource_helper.resource_request(endpoint=endpoint, method="put", payload=payload)
+
```
Comment (cmcnally-r7) [inline plugins/rapid7_insightvm/komand_rapid7_insightvm/actions/update_shared_credential/action.py:28]: Loving it, it's so neat and tidy now

File: plugins/rapid7_insightvm/komand_rapid7_insightvm/util/update_shared_credential_util.py [hunk 14]
```diff
@@ -0,0 +1,207 @@
...
+    if permission_elevation not in ("none", "pbrun"):
+        permission_elevation_username = check_not_null(account, "permission_elevation_username")
+        permission_elevation_password = check_not_null(account, "permission_elevation_password")
+    else:
+        permission_elevation_username = ""
+        permission_elevation_password = ""  # nosec
+    return permission_elevation, permission_elevation_username, permission_elevation_password
+
+
+def usr_and_pass(account: dict):
```
Comment (cmcnally-r7) [inline plugins/rapid7_insightvm/komand_rapid7_insightvm/util/update_shared_credential_util.py:22]: Return type hint on this

File: plugins/rapid7_insightvm/komand_rapid7_insightvm/util/update_shared_credential_util.py [hunk 15]
```diff
@@ -0,0 +1,207 @@
+from typing import Union
```
Comment (cmcnally-r7) [inline plugins/rapid7_insightvm/komand_rapid7_insightvm/util/update_shared_credential_util.py:1]:
```suggestion
from typing import Union, Tuple
```

File: plugins/rapid7_insightvm/komand_rapid7_insightvm/util/update_shared_credential_util.py [hunk 16]
```diff
@@ -0,0 +1,207 @@
...
+
+from insightconnect_plugin_runtime.exceptions import PluginException
+
+from komand_rapid7_insightvm.util.util import check_in_enum, check_not_null
+
+
+def ssh_setup(account: dict) -> tuple[str, str, str]:
```
Comment (cmcnally-r7) [inline plugins/rapid7_insightvm/komand_rapid7_insightvm/util/update_shared_credential_util.py:8]:
```suggestion
def ssh_setup(account: dict) -> Tuple[str, str, str]:
```

File: plugins/rapid7_insightvm/komand_rapid7_insightvm/util/update_shared_credential_util.py [hunk 17]
```diff
@@ -0,0 +1,207 @@
...
+        permission_elevation_username = ""
+        permission_elevation_password = ""  # nosec
+    return permission_elevation, permission_elevation_username, permission_elevation_password
+
+
+def usr_and_pass(account: dict):
+    return check_not_null(account, "username"), check_not_null(account, "password")
+
+
+def as400_cifs_cvs(account: dict, service: str) -> dict:
```
Comment (cmcnally-r7) [inline plugins/rapid7_insightvm/komand_rapid7_insightvm/util/update_shared_credential_util.py:26]: Can we get docstrings on all these?

File: plugins/rapid7_insightvm/komand_rapid7_insightvm/util/util.py [hunk 18]
```diff
@@ -91,3 +92,31 @@ def read_from_cache(filename):
...
+    :return: value or PluginException
+    """
+    value = account.get(var_name)
+    if value in (None, ""):
+        raise PluginException(cause=f"{var_name} has not been entered.", assistance=f"Enter valid {var_name}")
+    else:
+        return value
+
+
+def check_in_enum(value: str, var_name: str, possible_enums: list) -> None:
```
Comment (cmcnally-r7) [inline plugins/rapid7_insightvm/komand_rapid7_insightvm/util/util.py:111]:
```suggestion
def check_if_enum(value: str, var_name: str, possible_enums: list) -> None:
```
Also, if we're checking enums then do we really need to check enums if IC already only gives the user certain options?

**Review by cmcnally-r7 — CHANGES_REQUESTED** [review]

File: plugins/rapid7_insightvm/komand_rapid7_insightvm/actions/update_shared_credential/action.py [hunk 19]
```diff
@@ -0,0 +1,29 @@
...
+        return {"links": response["links"]}
```
Comment (igorski-r7) [inline plugins/rapid7_insightvm/komand_rapid7_insightvm/actions/update_shared_credential/action.py:29]: Kindly use the `OUTPUT` from schema 😄 
```suggestion
        return {Output.LINKS: response.get("links")}
```

File: plugins/rapid7_insightvm/komand_rapid7_insightvm/util/update_shared_credential_util.py (hunk 16 above)
Comment (igorski-r7) [inline plugins/rapid7_insightvm/komand_rapid7_insightvm/util/update_shared_credential_util.py:8]: I am up for using the `Dict` from typing library 😄 Add there necessary structure types as well for example `Dict[str, Any]`
```suggestion
def ssh_setup(account: Dict[str, Any]) -> tuple[str, str, str]:
```

File: plugins/rapid7_insightvm/komand_rapid7_insightvm/util/update_shared_credential_util.py (hunk 14 above)
Comment (igorski-r7) [inline plugins/rapid7_insightvm/komand_rapid7_insightvm/util/update_shared_credential_util.py:22]: Kindly please the return typehint and change dict to `Dict` from typing with a proper structure types

File: plugins/rapid7_insightvm/komand_rapid7_insightvm/util/update_shared_credential_util.py (hunk 17 above)
Comment (igorski-r7) [inline plugins/rapid7_insightvm/komand_rapid7_insightvm/util/update_shared_credential_util.py:26]: Please change `dict` to `Dict` with structured types 😄

File: plugins/rapid7_insightvm/komand_rapid7_insightvm/util/util.py [hunk 20]
```diff
@@ -91,3 +92,31 @@ def read_from_cache(filename):
         contents = cache_file.read()
 
         return contents
+
+
+def check_not_null(account: dict, var_name: str) -> Union[str, PluginException]:
```
Comment (igorski-r7) [inline plugins/rapid7_insightvm/komand_rapid7_insightvm/util/util.py:97]: Same here
```suggestion
def check_not_null(account: Dict[str, Any], var_name: str) -> Union[str]:
```

**Review by igorski-r7 — CHANGES_REQUESTED** [review]
Left some small comments related to typing 😄

File: plugins/rapid7_insightvm/komand_rapid7_insightvm/util/update_shared_credential_util.py (hunk 17 above)
Comment (TKuzas-R7) [reply plugins/rapid7_insightvm/komand_rapid7_insightvm/util/update_shared_credential_util.py:26]: 😭

**Review by hpwilson-r7 — APPROVED** [review]
Looks good 👍

File: plugins/rapid7_insightvm/komand_rapid7_insightvm/util/update_shared_credential_util.py [hunk 21]
```diff
@@ -0,0 +1,296 @@
+from typing import Union, Tuple, Dict
+
+from insightconnect_plugin_runtime.exceptions import PluginException
+
+from komand_rapid7_insightvm.util.util import check_not_null
+
+
+def ssh_setup(account: Dict[str, any]) -> Tuple[str, str, str]:
```
Comment (igorski-r7) [inline plugins/rapid7_insightvm/komand_rapid7_insightvm/util/update_shared_credential_util.py:8]: I suggest changing `any` to `Any` from `typing` library
 😄 
```suggestion
def ssh_setup(account: Dict[str, Any]) -> Tuple[str, str, str]:
```

**Comment by TKuzas-R7** [comment]
`INFO: RunController: Running /Users/tkuzas/dev/insightconnect-plugins/plugins/rapid7_insightvm/tests/update_shared_credential.json
INFO: RunController: Running command: docker run --rm -i rapid7/rapid7_insightvm:5.1.0 run < /Users/tkuzas/dev/insightconnect-plugins/plugins/rapid7_insightvm/tests/update_shared_credential.json
rapid7/Rapid7 InsightVM Console:5.1.0. Step name: update_shared_credential
Starting new HTTPS connection (1): ivm-console-test.vuln.lax.rapid7.com:3780
https://ivm-console-test.vuln.lax.rapid7.com:3780 "PUT /api/3/shared_credentials/3 HTTP/1.1" 200 None
Output is: {'body': {'log': 'rapid7/Rapid7 InsightVM Console:5.1.0. Step name: update_shared_credential\n', 'status': 'ok', 'meta': {}, 'output': {'links': [{'href': 'https://ivm-console-test.vuln.lax.rapid7.com:3780/api/3/shared_credentials/3', 'rel': 'self'}]}}, 'version': 'v1', 'type': 'action_event'}

{"body": {"log": "rapid7/Rapid7 InsightVM Console:5.1.0. Step name: update_shared_credential\n", "status": "ok", "meta": {}, "output": {"links": [{"href": "https://ivm-console-test.vuln.lax.rapid7.com:3780/api/3/shared_credentials/3", "rel": "self"}]}}, "version": "v1", "type": "action_event"}
`

**Review by cmcnally-r7 — APPROVED** [review]

**Review by llaszuk-r7 — APPROVED** [review]

**Pull request merged**